### PR TITLE
core: framework to register PM callbacks

### PR DESCRIPTION
**Pull request by etienne-lms**
This RFC is related to P-R https://github.com/OP-TEE/optee_os/pull/2683.
This proposes a framework for runtime registration of PM callback services.

Description (commit comment)
> Introduce a framework for power management callback registering.
Any core component can register a callback function that shall be
called with a cookie argument for suspend/resume sequence.
>
> Callbacks are related to a callback level. It defines the callbacks
call ordering, allowing core low level drivers (as clocks or the GIC)
to be suspended after all drivers and resume before these.
>
> Callbacks are related to a target subsystem identified with an
opaque ID. When suspending (or resuming) a target subsystem, only the
related callbacks are called. It is up to the platform to register
callbacks with appropriate subsystem IDs and to call suspend/resume
sequences for the target subsystems.

### CODE REVIEW

File: core/include/kernel/generic_pm.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * Callback shall return 0 on sucess and return 1 in case of failure. This
+ * may be used during a failing suspend sequence to call subsystem resume
+ * sequence.
+ *
+ * @pm_subsys defines the identifier which is used to notify generic PM
+ * of a subsystem suspend/resume. Callback registered for that subsystem
+ * will be suspended/resumed. Subsystem ID must be lower than UINT16_MAX.
+ */
+void register_pm_cb2(int (*callback)(enum pm_op op, void *handle),
```
Comment (jenswi-linaro) [inline core/include/kernel/generic_pm.h]: I'd prefer `register_pm_cb()`

File: core/include/kernel/generic_pm.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		     void *handle, enum pm_cb_level cb_level,
+		     unsigned int subsys);
+
+/*
+ * register_pm_cb - Register a driver callback for generic suspend/resume
+ *
+ * @callback: Registered callback function
+ * @handle: Registered cookie used as callback argument
+ */
+static inline void register_pm_cb(int (*callback)(enum pm_op op, void *handle),
```
Comment (jenswi-linaro) [inline core/include/kernel/generic_pm.h]: `register_pm_driver_cb()`

File: core/include/kernel/generic_pm.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * pm_suspend_resume - Call registered suspend/resume callbacks
+ *
+ * In case of failure when suspending, platform can resume the target
+ * subsystem, only the still suspended components will be resumed.
+ *
+ * @pm_subsys: Target subsystem suspended or resumed or 0 if none
+ * @op: Either PM_OP_SUSPEND or PM_OP_RESUME
+ * Return 0 on success, 1 on failure of a callback.
+ */
+int pm_suspend_resume(unsigned int subsys, enum pm_op op);
```
Comment (jenswi-linaro) [inline core/include/kernel/generic_pm.h]: `pm_change_state()`
Also I wonder if we shouldn't return a `TEE_Result` instead.

File: core/kernel/generic_pm.c [hunk 4]
```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <keep.h>
+#include <kernel/generic_pm.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
```
Comment (jenswi-linaro) [inline core/kernel/generic_pm.c]: This isn't needed since `<types_ext.h>` is included.

File: core/kernel/generic_pm.c [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+ */
+
+static bool __maybe_unused cb_is_valid(int (*cb)(enum pm_op op, void *hdl),
+				       void *hdl)
+{
+	paddr_t cb_phy = virt_to_phys(cb);
+	paddr_t hdl_phy = virt_to_phys(hdl);
+	bool __maybe_unused valid;
+
+	valid = (phys_to_virt(cb_phy, MEM_AREA_TEE_RAM_RX) == cb) &&
```
Comment (jenswi-linaro) [inline core/kernel/generic_pm.c]: We should have a global `is_unpaged()` helper function instead, and that function should compare the virtual address with suitable symbols from the links script instead.

File: core/kernel/generic_pm.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	}
+}
+
+static bool __maybe_unused pm_subsys_is_valid(unsigned int pm_subsys)
+{
+	return pm_subsys <= UINT16_MAX;
+}
+
+struct pm_cb {
+	int (*callback)(enum pm_op op, void *handle);
```
Comment (jenswi-linaro) [inline core/kernel/generic_pm.c]: The callback should probably get all the other elements of this struct as arguments.

File: core/kernel/generic_pm.c [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+	       pm_subsys_is_valid(pm_subsys));
+
+	size_t count = pm_cb_count;
+	struct pm_cb *pm_cb = pm_cb_ref;
+
+	pm_cb = realloc(pm_cb, sizeof(struct pm_cb) * (count + 1));
+	if (!pm_cb)
+		panic();
+
+	memset(&pm_cb[count], 0, sizeof(struct pm_cb));
```
Comment (jenswi-linaro) [inline core/kernel/generic_pm.c]: How about writing this as:
```
pm_cb[count] = (struct pm_cb){.callback = callback, .handle = handle,
                              .level = cb_level, .subsys = pm_subsys, };
```
We haven't used this pattern much, but it would provide better type safety and is a bit more compact while quite readable. Also note that I skipped the casts as they aren't needed.

**Comment by etienne-lms** [comment]
@cneveux: ok I will remove the `subsys` identifier and move to a suspend level based trigger.

Based on a suspend level (or power state) I see an issue. Platform have each their own power states. Specific IDs for specific meaning. Using such value in a driver would bind a driver to a specific platform which is not a good idea. How to be generic this way?

File: core/kernel/generic_pm.c (hunk 5 above)
Comment (etienne-lms) [reply core/kernel/generic_pm.c]: ok, i will look at that.
Regarding the tested value, I intentionally considered secure RAMs area as invalid candidate. Is that OK to you?

File: core/kernel/generic_pm.c (hunk 6 above)
Comment (etienne-lms) [reply core/kernel/generic_pm.c]: yes probably. I think the callback proto will even change since comments from @cneveux.

File: core/kernel/generic_pm.c (hunk 7 above)
Comment (etienne-lms) [reply core/kernel/generic_pm.c]: right.

**Comment by cneveux** [comment]
@etienne-lms 
This is where difficulty is.
In my implementation, I used a default PM State that is overwritten in platform_pwr.h. As the PCSI suspend function is specific to platform, it's the function that give the PM State to enter to the generic PM api that call the driver's callback (giving in parameter to state to enter).

In case of generic driver, it can support the default modes (that must be also defined in the specific platform enumerate). 
Other drivers are specific to the plaform so no issue.

File: core/kernel/generic_pm.c (hunk 5 above)
Comment (jenswi-linaro) [reply core/kernel/generic_pm.c]: I guess so. Propose something and we'll take it from there.

**Comment by etienne-lms** [comment]
A platform drivers shared between platform should be able to benefit from multi-domain suspend.
I think going this would mandate share generic IDs. I fear a bit how to handle that and help platforms power topology to use such schemes.

I guess my proposal for subsystems (could be clocks, power, bus access) would meet the same issue: binding platform topology to devices power description.

**Comment by etienne-lms** [comment]
Since file name changes and many fixes in the function prototypes, I've wiped the change history and pushed as 3 commits.
@cneveux, the 3rd commit of the series adds suspend state  hints the platform could share with the system. Would that match your needs?

File: core/arch/arm/mm/core_mmu.c [hunk 8]
```diff
@@ -1944,3 +1944,21 @@ vaddr_t core_mmu_get_va(paddr_t pa, enum teecore_memtypes type)
 
 	return (vaddr_t)pa;
 }
+
+bool is_unpaged_code(void *va)
+{
+	paddr_t pa = virt_to_phys(va);
+
+	return pa && phys_to_virt(pa, MEM_AREA_TEE_RAM_RX) == va;
```
Comment (jenswi-linaro) [inline core/arch/arm/mm/core_mmu.c]: This isn't accurate since `phys_to_virt()` will use `phys_to_virt_tee_ram()` for the lookup.

I don't see how it's necessary for debug purposes to check if something resides in data or code section, it seems you're trying to catch a problem that only can occur if created deliberately. However, I agree that it quite friendly to check that the function/data resides in the unpaged section as it's quite easy to get that wrong by mistake.

So checking that the virtual address is in the the range TEE_LOAD_ADDR..get_linear_map_end() in the page case should be enough.

File: core/include/kernel/pm.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+#include <stdbool.h>
+#include <stdint.h>
+
+/*
+ * Platform hints on targeted power state. Lower bits
+ * are generic resource bit flags. Higher bits are platform
+ * specific. Registered callbacks may choose to use or ignore
+ * these hints.
+ *
+ * PM_STATE_HINT_CLOCK - When set clock shall be suspended/restored
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/kernel/pm.h]: Fine with me. We should keep in mind that there's no need for a stable ABI here.

File: core/include/kernel/pm.h [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+enum pm_op {
+	PM_OP_SUSPEND = 0,
+	PM_OP_RESUME = 1,
+};
+
+/*
+ * Registered callbacks are called the ordering directives
+ * specified by the PM_CB_LEVEL_xxx value.
+ */
+enum pm_cb_level {
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: Since this is a pure ordering thing I'd suggest `enum pm_callback_order`, it's fine with `PM_CB_ORDER_` prefix for the values.

File: core/include/kernel/pm.h [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+};
+
+#define PM_CB_HDL(_cb, _hdl, _level)		\
+		(struct pm_cb_hdl){		\
+			.callback = (_cb),	\
+			.handle = (_hdl),	\
+			.level = (_level),	\
+		}
+
+#define PM_CB_ARG_HANDLE(pm_cb_hdl)		((pm_cb_hdl)->handle)
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: Is this really needed?

File: core/include/kernel/pm.h [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+ * Registered callbacks are called the ordering directives
+ * specified by the PM_CB_LEVEL_xxx value.
+ */
+enum pm_cb_level {
+	PM_CB_LEVEL_DRIVER = 0,
+	PM_CB_LEVEL_CORE_SERVICE,
+	PM_CB_LEVELS
+};
+
+#define PM_CB_HDL(_cb, _hdl, _level)		\
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: I'd prefer spelling out most of the abbreviations as this looks a bit too cryptic.
How about PM_CALLBACK_HANDLE_INITIALIZER()?

File: core/include/kernel/pm.h [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+		}
+
+#define PM_CB_ARG_HANDLE(pm_cb_hdl)		((pm_cb_hdl)->handle)
+
+/*
+ * @callback - Registered callback function
+ * @handle - Registered private handler for the callback
+ * @level - Registered callback call order priority (PM_CB_PRIO_*)
+ */
+struct pm_cb_hdl {
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: I'd prefer `struct pm_callback_handle`

File: core/include/kernel/pm.h [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+	/* Set by the caller when registering a callback */
+	int (*callback)(enum pm_op op, unsigned int pm_state,
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: Please document the arguments for the callback. What's `pm_state` for instance?

File: core/include/kernel/pm.h [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+
+/*
+ * Request call to registered PM callbacks
+ *
+ * @op: Either PM_OP_SUSPEND or PM_OP_RESUME
+ * @pm_state: Platform power state we are suspending to or resuming from
+ *
+ * Return 0 on success, 1 on failure of a callback.
+ */
+int pm_change_state(enum pm_op op, unsigned int pm_state);
```
Comment (jenswi-linaro) [inline core/include/kernel/pm.h]: So now we have yet another set of return codes. Until this we had:
bool (true, false)
int (0 success, < 0 failure)
TEE_Result (TEE_SUCCES/0 success, anything else error)

Unless there's a good reason I think we should stick to TEE_Result just as the Linux kernel normally uses -errno.

Comment (jbech-linaro) [reply core/include/kernel/pm.h]: +1, since having various return codes is a pain in the end. We're already enforced to convert error codes to non TEE error codes in many cases on non-secure side when returning back from the TEE (KM/GK, Widevine, PKCS#11 to name a few). So trying to only use TEE_Result on secure side seems like a good idea.

**Comment by etienne-lms** [comment]
Comments addressed. I tested this PM implementation on my platform. Works OK.

File: core/include/kernel/pm.h (hunk 11 above)
Comment (etienne-lms) [reply core/include/kernel/pm.h]: I preserved the macro, renamed into `PM_CALLBACK_GET_HANDLE()`.

File: core/kernel/pm.c [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+				return res;
+		}
+		break;
+	default:
+		panic();
+	}
+
+	return TEE_SUCCESS;
+}
+KEEP_PAGER(pm_change_state);
```
Comment (jenswi-linaro) [inline core/kernel/pm.c]: Why is this needed? If this function is called from unpaged code it should automatically end up in the unpaged area.

**Comment by jenswi-linaro** [comment]
For the commit "core: introduce is_unpaged() function
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`":

File: core/include/kernel/pm.h [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+/*
+ * Register a driver callback for generic suspend/resume.
+ * Refer to struct pm_callback_handle for description of the callbacks
+ * API.
+ *
+ * @callback: Registered callback function
+ * @handle: Registered private handle argument for the callback
+ */
+void register_pm_driver_cb(
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/kernel/pm.h]: Ok, I will make it as an inline function. 
As a driver writer, I would prefer to implementing something like 
```c
	(...)
	register_pm_driver_cb(gpio_secure_suspend_resume, dev);
	(...)
```
rather than
```c
	(...)
	register_pm_cb(&PM_CALLBACK_HANDLE_INITIALIZER(
				gpio_secure_suspend_resume, dev,
				PM_CB_ORDER_DRIVER));
	(...)
```

File: core/kernel/pm.c (hunk 16 above)
Comment (etienne-lms) [reply core/kernel/pm.c]: yes, that's right. I will remove this `KEEP_PAGER()`.

Maybe I should change `verify_cb_args()`:
```c
	if (is_unpaged((void *)(vaddr_t)pm_change_state) &&
	    (!is_unpaged((void *)(vaddr_t)pm_hdl->callback) ||
	     (pm_hdl->handle && !is_unpaged(pm_hdl->handle))) {
		//error
	}
```

File: core/include/kernel/pm.h [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+ * PM_OP_RESUME: platform is resuming from low power state
+ */
+enum pm_op {
+	PM_OP_SUSPEND = 0,
+	PM_OP_RESUME = 1,
+};
+
+/*
+ * Registered callbacks are called the ordering directives
+ * specified by the PM_CB_LEVEL_xxx value.
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/include/kernel/pm.h]: thanks

File: core/kernel/pm.c (hunk 16 above)
Comment (jenswi-linaro) [reply core/kernel/pm.c]: Makes sense, could save a lot of troubles. :-)

File: core/include/kernel/pm.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/include/kernel/pm.h]: Fair enough.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
tags applied

**Comment by cneveux** [comment]
@etienne-lms I verified on my side and it working too. Thanks

**Comment by etienne-lms** [comment]
Thanks a lot @cneveux, great. I will not apply your `tested-by` tag or mine since this cannot be tested from pristine source tree.
I will push an update with the minor changes I've commented.

**Comment by etienne-lms** [comment]
Late renamed `PM_STATE_HINT_xxx` into `PM_HINT_xxx_STATE` and updated arg label in function prototype. 
@cneveux, maybe I can apply your `Reviewed-by` tag on the 2nd commit?

**Comment by cneveux** [comment]
@etienne-lms 
Do as you want , it's not important for me. I just wanted to inform you that it's ok for me.

Thanks

**Comment by etienne-lms** [comment]
ok. thanks for contributions.
ready to be merged.

**Comment by jforissier** [comment]
I have added @cneveux's `Acked-by:` to the second patch since I think it best reflects the discussion. PR merged, thanks.
